certbot/ocsp.py

```diff
@@ -207,11 +207,43 @@ def _check_ocsp_response(response_ocsp, request_ocsp, issuer_cert):
         raise AssertionError('param nextUpdate is in the past.')
 
 
-def _check_ocsp_response_signature(response_ocsp, issuer_cert):
-    """Verify an OCSP response signature against certificate issuer"""
+def _check_ocsp_response_signature(response_ocsp, issuer_cert, cert_path):
+    """Verify an OCSP response signature against certificate issuer or responder"""
+    if response_ocsp.responder_name == issuer_cert.subject:
+        logger.debug('OCSP response for certificate %s is signed by the certificate\'s issuer.',
+                     cert_path)
+        responder_cert = issuer_cert
+    else:
+        logger.debug('OCSP response for certificate %s is delegated to an external responder.',
+                     cert_path)
+
+        responder_certs = [cert for cert in response_ocsp.certificates
+                           if cert.subject == response_ocsp.responder_name]
+        if not responder_certs:
+            raise AssertionError('no matching responder certificate could be found')
+
+        responder_cert = responder_certs[0]
+
+        if responder_cert.issuer != issuer_cert.subject:
+            raise AssertionError('responder certificate is not signed '
+                                 'by the certificate\'s issuer')
+
+        try:
+            extension = responder_cert.extensions.get_extension_for_class(x509.ExtendedKeyUsage)
+            authorized = x509.oid.ExtendedKeyUsageOID.OCSP_SIGNING in extension.value
```

Comment:
nit: I was confused about this variable name for a while. Maybe `delegated` would fit better :)

Reply:
I merge the two, because it is not only that it is a delegate, it needs also to be explicitly authorized. So `delegate_authorized`.